frame/dex-router/src/lib.rs

```diff
@@ -242,43 +402,117 @@ pub mod pallet {
 				)?;
 				dx_t = dy_t;
 			}
+			ensure!(dy_t >= min_receive, Error::<T>::CanNotRespectMinAmountRequested);
 			Ok(dy_t)
 		}
 
+		#[transactional]
 		fn sell(
 			who: &T::AccountId,
 			asset_pair: CurrencyPair<T::AssetId>,
 			amount: T::Balance,
+			min_receive: T::Balance,
 		) -> Result<T::Balance, DispatchError> {
-			Self::exchange(who, asset_pair, amount)
+			<Self as DexRouter<
+				T::AccountId,
+				T::AssetId,
+				T::PoolId,
+				T::Balance,
+				T::MaxHopsInRoute,
+			>>::exchange(who, asset_pair, amount, min_receive)
 		}
 
+		#[transactional]
 		fn buy(
 			who: &T::AccountId,
 			asset_pair: CurrencyPair<T::AssetId>,
 			amount: T::Balance,
+			min_receive: T::Balance,
 		) -> Result<T::Balance, DispatchError> {
-			let route = Self::get_route(asset_pair).ok_or(Error::<T>::NoRouteFound)?;
+			let (route, reverse) = Self::get_route(asset_pair).ok_or(Error::<T>::NoRouteFound)?;
 			let mut dy_t = amount;
 			let mut dx_t = T::Balance::zero();
-			for pool_id in route.iter().rev() {
-				let currency_pair = T::Pablo::currency_pair(*pool_id)?;
+			let mut forward_iter;
+			let mut backward_iter;
+			let route_iter: &mut dyn Iterator<Item = &T::PoolId> = if !reverse {
+				backward_iter = route.iter().rev();
+				&mut backward_iter
+			} else {
+				forward_iter = route.iter();
+				&mut forward_iter
+			};
+			for pool_id in route_iter {
+				let mut currency_pair = T::Pablo::currency_pair(*pool_id)?;
+				if reverse {
+					currency_pair = currency_pair.swap();
+				}
 				dx_t = T::Pablo::get_exchange_value(*pool_id, currency_pair.base, dy_t)?;
 				dy_t = dx_t;
 			}
-			for pool_id in route {
-				let currency_pair = T::Pablo::currency_pair(pool_id)?;
+			let route_iter: &mut dyn Iterator<Item = &T::PoolId> = if !reverse {
+				forward_iter = route.iter();
+				&mut forward_iter
+			} else {
+				backward_iter = route.iter().rev();
+				&mut backward_iter
+			};
+			for pool_id in route_iter {
+				let mut currency_pair = T::Pablo::currency_pair(*pool_id)?;
+				if reverse {
+					currency_pair = currency_pair.swap();
+				}
 				let dy_t = T::Pablo::exchange(
 					who,
-					pool_id,
+					*pool_id,
 					currency_pair,
 					dx_t,
 					T::Balance::zero(),
 					true,
 				)?;
 				dx_t = dy_t;
 			}
+			ensure!(dx_t >= min_receive, Error::<T>::CanNotRespectMinAmountRequested);
 			Ok(dx_t)
 		}
+
+		#[transactional]
+		fn add_liquidity(
+			who: &T::AccountId,
+			asset_pair: CurrencyPair<T::AssetId>,
+			base_amount: T::Balance,
+			quote_amount: T::Balance,
+			min_mint_amount: T::Balance,
+			keep_alive: bool,
+		) -> Result<(), DispatchError> {
+			let (route, _reverse) = Self::get_route(asset_pair).ok_or(Error::<T>::NoRouteFound)?;
+			if route.len() != 1 {
```

Comment:
why not get pair directly here?

Reply:
if route gets updated and still remains signle node route it will work fine, otherwise it will be an UnsupportedOperation error.